nidirect_workflow/nidirect_workflow.module

```diff
@@ -11,3 +20,59 @@ function nidirect_workflow_page_attachments(array &$attachments) {
   // Attach extra custom css for admin menu.
   $attachments['#attached']['library'][] = 'nidirect_workflow/admin.css';
 }
+
+/**
+ * Implements hook_entity_presave().
+ */
+function nidirect_workflow_entity_presave(EntityInterface $entity) {
+  // This will fire when nodes are created or edited.
+  if ($entity->getEntityTypeId() == 'node') {
+    switch ($entity->bundle()) {
+      case 'article':
+      case 'contact':
+      case 'page':
+        // We have just published a node type that is subject to auditing.
+        if ($entity->isPublished()) {
+          $next_audit_date = $entity->get('field_next_audit_due')->value;
+          if (empty($next_audit_date)) {
+            // No audit date set, set it for six months time.
+            $entity->set('field_next_audit_due', date('Y-m-d', strtotime("+6 months")));
+          }
+        }
+        break;
+    }
+  }
+}
+
+/**
+ * Implements hook_form_alter().
+ */
+function nidirect_workflow_form_alter(&$form, FormStateInterface $form_state, $form_id) {
+
+  // Manage access to 'next audit date' field.
+  if (($form_id == 'node_article_form')
```

Comment:
I'd pop these into an array and run a single: `if (in_array($values, $form_id) { .. }` or alternatively `if (preg_match('/pattern/', $form_id) { .. }` for brevity.

Caveat: I can never remember the order of parameters in `in_array()` :D

Reply:
Personally I prefer it as is - I think it is more readable !

Reply:
I'd be more of a fan of in_array()
Happy medium could be:

  if (in_array($form_id, [
    'node_article_edit_form',
    'node_contact_form',
    'node_contact_edit_form'])) {

  }

Reply:
OK, will go with that